Style 2021 (#389)
VisionSystemsInc/vsi_common: docs/style.rst

  .. code-block:: bash

     run_some_command &> /dev/null # Spaces added around &>

     run_some_command >& /dev/null # Wrong style. This is dealing with filenames

* Always quote variables and complex strings

  There are many conditions when you need to use quotes, and very few when you need not to, and many when quotes seem optional. Always using quotes vastly reduces the chance of errors from unexpected (and erroneous) input and simplifies remembering when you need and don't need quotes

  .. code-block:: bash

     bvar=${avar}   # Wrong style. Always add quotes
     bvar="${avar}" # Right
     cvar=foo bar   # Wrong style and syntax. This is a complex string, and requires quotes/
                    # This actually executes a command called bar. Instead, make it clear
     cvar="foo" bar # Right
     cvar="foo bar" # Right
     echo ${avar}   # Wrong style. Whitespace is not replicated the same way without quotes
     echo "${avar}"

     # There are a few reason to not quote a variable. Be sure to add a "# noquotes"
     # comment so code reviewers knows these are intended:

     # One is an empty variable that should not be treated as an empty string

>      # DRYRUN/optional_flag could be "echo"/"-stuff" or empty string

Reviewer: ```suggestion
     # E.g., DRYRUN could be "echo" or empty string
```

     ${DRYRUN} some command # noquotes

>      some_command ${optional_flag} foo bar # noquotes

Reviewer: ```suggestion
     # E.g., optional_flag could be "-stuff" or empty string
     some_command ${optional_flag} foo bar # noquotes
```

     # The preferred way is to use arrays, if it is not overly cumbersome.
     ${DRYRUN[@]+"${DRYRUN[@]}"} some command
     some_command ${optional_flag[@]+"${optional_flag[@]}"} foo bar

>      # Since arrays cannot be exported, this is often not viable.

Reviewer: ```suggestion
     # Although because arrays cannot be exported, this is often not viable.
```

>      # Another reason quotes must not be used, is when splitting up a string into an array

Reviewer: ```suggestion
     # Another reason quotes must not be used is when splitting up a string into an array
```

     foo="aa:bb:cc:dd"
     IFS=":"
     # This is actively splitting apart a string, and must not be in quotes

>      bar=(${foo}) # noquotes

Reviewer: Should this be `bar=(${foo[@]})`

Author: No, foo is a string, not an array. bar is the array

>   Quotes should actually not be used in ``[[]]`` expressions. There are a few corner cases the will be treated literally. ``# noquotes`` is not needed for ``[[]]`` expressions.

Reviewer: ```suggestion
  Quotes should not be used in ``[[]]`` expressions---there are a few corner cases where the quotes will be treated literally. ``# noquotes`` is not needed for ``[[]]`` expressions.
```

       echo "hi"
     fi

  Simple assignments _may_ skip quotes, ``# noquotes`` is not needed as this is easy enough for a code reviewer to see

  .. code-block:: bash

     local a=1
     b=2
     x=(11 22 33 44 "5 5" "6  6")
     cvar=foo
     dvar=foo\ bar  # Wrong style. This is no longer simple. Use quotes
     dvar="foo bar" # Right

  It is best not to use quotes when inside of ``{}``. The expressions inside of ``{}`` can be thought of as already being implicitly quoted (``"``). Adding quotes (``'`` or ``"``) may seem to work at first, but the behavior of explicitly quoting will change between the different versions of bash.

  .. code-block:: bash

     echo "${foo-"b a  r"}"        # Wrong style
     echo "${foo-b a  r}"          # Right
     echo "${foo/"o"/"O"}"         # Wrong style
     echo "${foo/o/O}"             # Right
     echo "${foo/"  "/two spaces}" # Wrong style
     echo "${foo/  /two spaces}"   # Right

  :var:`bash_behavior_pattern_substitution_slash_escape_with_single_quote` is a special case that still needs special care, due to differences in bash behavior between versions.

* Always use ${var} vs $var

>   The reason for this policy is consistency and to clarify that certain features in bash only work in the ``{}``, e.g. variable substitution. It's very easy for someone to mistake ``${foo+set}`` for ``$foo+set`` and not ``${foo}+set``.

Reviewer: is this backwards? should it be, "It's very easy for someone to mistake ``$foo+set`` for ``${foo}+set`` and not ``${foo+set}`` or something?

>   The reason for this policy is consistency and to clarify that certain features in bash only work in the ``{}``, e.g. variable substitution. It's very easy for someone to mistake ``${foo+set}`` for ``$foo+set`` and not ``${foo}+set``.

Author: ```suggestion
  The reason for this policy is consistency and to clarify that certain features in bash only work in the ``{}``, e.g. variable substitution. It's very easy for someone to mistake ``$foo+set`` for ``${foo}+set`` and not ``${foo+set}``.
```

  .. code-block:: bash

     echo "$PATH"                   # Wrong style
     echo "${PATH}"                 # Right
     echo "${$}"                    # Right

>      echo "${-} ${?} ${*-}"         # * and @ need some extra care, so that

Reviewer: so that what?

     run command "${_}" ${@+"${@}"} # set -eu doesn't error on empty in bash 3.2

* Shorthand for arithmetic expressions

  .. code-block:: bash

     x=(11 22 33 44)
     y=2
     echo "${x[y]} is perfectly acceptable"

>      echo "${x[$y]} is violated the {} policy, even though it is valid bash"

Reviewer: ```suggestion
     echo "${x[$y]} violates the {} policy, even though it is valid bash"
```

     echo "${x[${y}]} is ok too, but the shorthand looks better"
     echo "$((x[y] - y)) is also perfectly acceptable"
     echo "${x:1:y} is also perfectly acceptable"
     echo "${x:1:y+1} is also perfectly acceptable"

     # Do no add quotes to inner expressions

>      echo "${x["y"]} ${x["${y}"]}"

Reviewer: ```suggestion
     echo "${x["y"]} ${x["${y}"]}" # Wrong style
```

     # Associative arrays are not bash 3.2 compatible, and are not

>      # arithmetic expressions in the []

Reviewer: ```suggestion
     # arithmetic expressions in []
```
